### PR TITLE
URL Pattern Match and Rewrite - Add extension structure and layout

### PR DESCRIPTION
**Pull request by silverstar194**
Commit Message:
URL Pattern Match and Rewrite - Add extension structure and layout #22459

Additional Description:
Added extension layout and setup for implementation of url pattern matching and rewrite.

The change are outlined as follows:
PathMatchPredicate - Added as a interface for path match extensions
PathMatchPredicateFactory - Added to generate path match extension instances 
PatternTemplateMatchPredicateFactory  - Generate specific pattern match extension

PathRewritePredicate - Added as a interface for path rewrite extensions
PathRewritePredicateFactory  - Added to generate path rewrite extension instances
PatternTemplateRewritePredicateFactory - Generate specific pattern rewrite extension

Full set of changes to be merged is mostly complete and tested. It does still need cleaning up. I'll be splitting PRs off of it here:
https://github.com/envoyproxy/envoy/pull/22207

Risk Level:
No changes to existing functionality within this PR.

Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -265,6 +265,10 @@ extensions/filters/http/oauth2 @derekargueta @snowp
 /*/extensions/filters/http/header_to_metadata @zuercher @JuniorHsu
 # zookeeper
 /*/extensions/filters/network/zookeeper_proxy @snowp @JuniorHsu
+# path match by pattern
+/*/extensions/path/match/pattern_template @silverstar194
+# path rewrite by pattern
+/*/extensions/path/rewrite/pattern_template @silverstar194
```
Comment (alyssawilk) [inline CODEOWNERS:271]: you can tag me as second owner

File: envoy/router/path_match_policy.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  /**
+   * @return the name of the current predicate.
+   */
+  virtual std::string name() const PURE;
+
+  /**
+   * Used to determine if the current url matches the predicate pattern.
+   *
+   * @param pattern current url of route
```
Comment (alyssawilk) [inline envoy/router/path_match_policy.h:30]: is it the literal url or a regex pattern?  let's make sure naming makes sense

File: envoy/router/path_match_policy.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   * @return vaild is route url matches the precidate pattern.
```
Comment (alyssawilk) [inline envoy/router/path_match_policy.h:31]: is -> if?

File: envoy/router/path_rewrite_policy.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  /**
+   * @return the name of the current predicate.
+   */
+  virtual absl::string_view name() const PURE;
+
+  /**
+   * Used to rewrite the current url to the specified output. Can return a failure in case rewrite
+   * is not successful.
+   *
+   * @param current_pattern current url of route
```
Comment (alyssawilk) [inline envoy/router/path_rewrite_policy.h:31]: again if it's the url, maybe call it that?

File: CODEOWNERS [hunk 5]
```diff
@@ -265,6 +265,10 @@ extensions/filters/http/oauth2 @derekargueta @snowp
 /*/extensions/filters/http/header_to_metadata @zuercher @JuniorHsu
 # zookeeper
 /*/extensions/filters/network/zookeeper_proxy @snowp @JuniorHsu
+# path match by pattern
```
Comment (silverstar194) [inline CODEOWNERS:268]: Who should I have as code owners here?

**Review by alyssawilk — COMMENTED** [review]
I think I'm not going to be able to review this without a feel for how it's implemented.  I assume that's one of the other PRs in flight?
/wait on CI, and use case PR :-)

**Comment by silverstar194** [comment]
> I think I'm not going to be able to review this without a feel for how it's implemented. I assume that's one of the other PRs in flight?

The full set of changes for context is here. Also in PR description with some more details around it. 
https://github.com/envoyproxy/envoy/pull/22207

File: CODEOWNERS (hunk 1 above)
Comment (silverstar194) [reply CODEOWNERS:271]: changed

File: envoy/router/path_rewrite_policy.h (hunk 4 above)
Comment (silverstar194) [reply envoy/router/path_rewrite_policy.h:31]: changed

File: envoy/router/path_match_policy.h (hunk 3 above)
Comment (silverstar194) [reply envoy/router/path_match_policy.h:31]: changed

File: envoy/router/path_match_policy.h (hunk 2 above)
Comment (silverstar194) [reply envoy/router/path_match_policy.h:30]: changed

**Comment by alyssawilk** [comment]
/wait on ci.
format is not going to pass.  please make sure format and clang tidy are passing and ping back?

**Comment by silverstar194** [comment]
Closing out for full PR: https://github.com/envoyproxy/envoy/pull/22207

**Pull request closed without merging**